### PR TITLE
os/bluestore: Improve _block_picker function

### PR DESCRIPTION
**Pull request by aclamk**
Make _block_picker function scan (*cursor, end) + (begin, *cursor) instead of (*cursor, end) + (begin, end).
The second run over range (*cursor, end) could never yield any results.

Signed-off-by: Adam Kupczyk <akupczyk@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by aclamk** [comment]
This PR is spinoff from #41369.
It just improves performance, does not fix problem addressed in #41369.

File: src/os/bluestore/AvlAllocator.cc [hunk 1]
```diff
@@ -45,14 +45,16 @@ uint64_t AvlAllocator::_block_picker(const Tree& t,
     }
   }
   /*
-   * If we know we've searched the whole tree (*cursor == 0), give up.
-   * Otherwise, reset the cursor to the beginning and try again.
+   * If we reached end, start from beginning till cursor.
    */
-   if (*cursor == 0 || rs_start == t.begin()) {
-     return -1ULL;
-   }
```
Comment (tchaikov) [inline src/os/bluestore/AvlAllocator.cc:53]: why is this part removed? i think it's still correct.

**Review by ifed01 — COMMENTED** [review]
Good!

**Review by ifed01 — APPROVED** [review]

File: src/os/bluestore/AvlAllocator.cc (hunk 1 above)
Comment (ifed01) [reply src/os/bluestore/AvlAllocator.cc:53]: +1

Comment (aclamk) [reply src/os/bluestore/AvlAllocator.cc:53]: It is removed, because if any of above conditions is true, then following loop is empty.
I removed it to simplify code.

Comment (ifed01) [reply src/os/bluestore/AvlAllocator.cc:53]: Just a remark here - when using best-fit mode (i.e. cursor=0 and lookup is performed using block size ordering) - one doesn't need to reiterate from the beginning - initial lookup has already provided the first suitable block - none are present before that position. Hence 
if(*cursor ==0) return -1;
might be useful.

**Review by ifed01 — APPROVED** [review]

**Review by tchaikov — APPROVED** [review]

**Comment by ifed01** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-25_04:17:20-rados-wip-kefu-testing-2021-05-25-1039-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-25_14:42:43-rados-wip-kefu-testing-2021-05-25-2008-distro-basic-smithi/

failure tracked by

- https://tracker.ceph.com/issues/45423
